### PR TITLE
[Core] Encode job ID in randomized task IDs for user-created threads

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

Currently, when `WorkerContext::GetCurrentTaskID()` returns a random task ID in user-created threads, and the returned task ID doesn't include the job ID. In this case, subsequent non-actor tasks and return values, and objects created by `ray.put()` don't include the job ID neither. This makes us hard to find the correct job ID from a task or object ID.

This PR updates the task ID generation code to always encode the job ID.

A side-effect of this PR is the change of possibility of task ID collision in user-created threads due to the fixed job ID part. w/o this PR: `sqrt(pi * 256 ^ 12 / 2)` ~= 352 trillion tasks. w/ this PR: `sqrt(pi * 256 ^ 8 / 2)` ~= 5 billion tasks. But this should be OK because the job ID part of task IDs in non-user-created threads are always fixed, so it won't be worse than non-user-created threads.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
I think this makes sense, but can you 

1. update https://github.com/ray-project/ray/blob/master/src/ray/design_docs/id_specification.md?
2. Fix broken tests?
3. What's the likelihood of collision with this mechanism? Can you calculate it?

**Comment by kfstorm** [comment]
@rkooo567 I don't think we need to update the ID spec doc since I didn't change the structure of task ID.

Task ID consists of 8 unique bytes and a job ID of 4 bytes. After this PR, the job ID part is fixed. For a 50% collision possibility, we need about 5 billion randomized task IDs for the same job, if don't count non-randomized task IDs.

```
>>> import math
>>> n = 256 ** 8
>>> n
18446744073709551616
>>> math.sqrt(math.pi * n / 2)
5382943231.384526
```

File: python/ray/includes/unique_ids.pxi [hunk 1]
```diff
@@ -171,8 +171,9 @@ cdef class TaskID(BaseID):
         return CTaskID.Size()
 
     @classmethod
-    def for_fake_task(cls):
-        return cls(CTaskID.ForFakeTask().Binary())
+    def for_fake_task(cls, job_id):
```
Comment (rkooo567) [inline python/ray/includes/unique_ids.pxi:174]: What does it mean by "fake task" here? The naming seems very confusing to me.

File: src/ray/core_worker/context.h [hunk 2]
```diff
@@ -107,7 +107,7 @@ class WorkerContext {
   mutable absl::Mutex mutex_;
 
  private:
-  static WorkerThreadContext &GetThreadContext();
+  WorkerThreadContext &GetThreadContext() const;
```
Comment (rkooo567) [inline src/ray/core_worker/context.h:110]: nice!

**Review by rkooo567 — APPROVED** [review]
I think it makes sense, but do you know what it means by "fake task id"? Is there a better wording than this? (I was really confused what it means when I read it.. )

**Comment by kfstorm** [comment]
@rkooo567 The "fake task id" here means a random task ID. It's used for user-created threads to avoid object/task ID collision when a user calls Ray APIs in multiple threads. Renaming `ForFakeTask` to `FromRandom` sounds reasonable to me, so `FromRandom()` and `FromRandom(const JobID &)` are overloads. Actually, the existing implementation of `TaskID::ForFakeTask` is exactly the same as `BaseID::FromRandom()`.

**Comment by scv119** [comment]
agreed ForFake is very confusing; let's rename it to FromRandom

**Comment by scv119** [comment]
Also, can you suggest the possibility of collision and consequence of it in the PR summary?

**Review by scv119 — CHANGES_REQUESTED** [review]
wait for author's response.

**Comment by rkooo567** [comment]
cc @scv119

**Review by scv119 — APPROVED** [review]

**Comment by kfstorm** [comment]
Windows CI failure is unrelated. Merging.

**Pull request merged**